sdk/opa.go

```diff
@@ -247,55 +300,80 @@ func (opa *OPA) Decision(ctx context.Context, options DecisionOptions) (*Decisio
-			now:             record.Timestamp,
-			path:            record.Path,
-			input:           *record.Input,
-			m:               record.Metrics,
-		})
-		if record.Error == nil {
-			record.Results = &result.Result
-		}
+		work(result)
 	}
 
 	m.Timer(metrics.SDKDecisionEval).Stop()
 
 	if logger := logs.Lookup(s.manager); logger != nil {
-		if err := logger.Log(ctx, &record); err != nil {
+		if err := logger.Log(ctx, record); err != nil {
 			return result, fmt.Errorf("decision log: %w", err)
 		}
 	}
-
-	return result, record.Error
+	return result, nil
 }
 
-// DecisionOptions contains parameters for query evaluation.
-type DecisionOptions struct {
-	Now   time.Time   // specifies wallclock time used for time.now_ns(), decision log timestamp, etc.
-	Path  string      // specifies name of policy decision to evaluate (e.g., example/allow)
-	Input interface{} // specifies value of the input document to evaluate policy with
-}
+// Partial returns a named decision. This function is threadsafe.
+func (opa *OPA) Partial(ctx context.Context, options PartialOptions) (*DecisionResult, error) {
 
-// DecisionResult contains the output of query evaluation.
-type DecisionResult struct {
-	ID     string      // provides a globally unique identifier for this decision (which is included in the decision log.)
-	Result interface{} // provides the output of query evaluation.
-}
+	if options.Mapper == nil {
+		return nil, fmt.Errorf("PartialQueryMapper must be specified")
+	}
 
-func newDecisionResult() (*DecisionResult, error) {
-	id, err := uuid.New(rand.Reader)
+	record := server.Info{
+		Timestamp: options.Now,
+		Input:     &options.Input,
+		Query:     options.Query,
+		Path:      options.Path,
+	}
+
+	result, err := opa.executeTransaction(
+		ctx,
+		&record,
+		func(result *DecisionResult) {
+			var pq *rego.PartialQueries
+			pq, record.InputAST, record.Bundles, record.Error = partial(ctx, partialEvalArgs{
+				runtime:   opa.state.manager.Info,
+				printHook: opa.state.manager.PrintHook(),
+				compiler:  opa.state.manager.GetCompiler(),
+				store:     opa.state.manager.Store,
+				txn:       record.Txn,
+				now:       record.Timestamp,
+				query:     record.Query,
+				unknowns:  options.Unknowns,
+				input:     *record.Input,
+				m:         record.Metrics,
+			})
+			if record.Error == nil {
+				var recordResult interface{}
+				result.Result, recordResult, record.Error = options.Mapper.MapResults(pq)
+				if record.Error == nil {
+					record.Results = &recordResult
+				}
+			}
+		},
+	)
 	if err != nil {
 		return nil, err
 	}
-	result := &DecisionResult{ID: id}
-	return result, nil
+
+	return result, record.Error
+}
+
+// The first interface being returned is the type that will be used for further processing
+// The second interface is for the decision logger, most of the time these should be the same
+type PartialQueryMapper interface {
+	MapResults(pq *rego.PartialQueries) (interface{}, interface{}, error)
```

Comment:
Let's change this to something like:


```golang
type PartialQueryMapper interface {
  MapResults(*rego.PartialQueries) (json.Marshaler, error)
}
```

The mapper will return an interface value that implements the MarshalJSON() function. This ensures that the SDK can marshal the result into the decision log record.

Reply:
I'm not sure how to make this work, with my limited knowledge of go, since I didn't define the type, I don't think I can add the function (e.g. I wouldn't be able to add the function to rego.PartialQueries in the RawMapper).  I could add the function directly to PartialQueryMapper, but then the type may be doing double work

Reply:
So instead of returning json.Marshaler, I decided to make a second function that specifically returns a type that can be logged as JSON.  Let me know if you think this is enough.